File: android/src/main/java/expo/modules/breathing/BreathingSharedState.kt
package expo.modules.breathing

enum class BreathPhase(val value: String) {
    INHALE("inhale"),
    HOLD_IN("holdIn"),
    EXHALE("exhale"),
    HOLD_OUT("holdOut"),
    IDLE("idle")
}

enum class BreathingExerciseState {
    STOPPED,
    RUNNING,
    PAUSED,
    COMPLETE
}

data class BreathPhaseConfig(
    val phase: BreathPhase,
    val duration: Double,  // seconds
    val targetScale: Double,
    val label: String
)

object BreathingSharedState {
    // Pattern configuration
    @Volatile
    var phases: List<BreathPhaseConfig> = emptyList()

    @Volatile
    var totalCycles: Int? = null  // null = infinite

    // Current state
    @Volatile
    var state: BreathingExerciseState = BreathingExerciseState.STOPPED

    @Volatile
    var currentPhaseIndex: Int = 0

    @Volatile
    var currentCycle: Int = 0

    @Volatile
    var phaseStartTime: Long = System.nanoTime()

    @Volatile
    var pauseTime: Long? = null

    // Animation values
    @Volatile
    var currentScale: Double = 1.0

    @Volatile
    var targetScale: Double = 1.0

    @Volatile
    var currentPhase: BreathPhase = BreathPhase.IDLE

    @Volatile
    var currentLabel: String = ""

    @Volatile
    var phaseProgress: Double = 0.0

    // Wobble animation
    @Volatile
    var wobbleIntensity: Double = 1.0

    @Volatile
    var wobbleOffsets: MutableList<Double> = mutableListOf()

    @Volatile
    var wobbleTargets: MutableList<Double> = mutableListOf()

    @Volatile
    var lastWobbleUpdate: Long = System.nanoTime()

    @Volatile
    var lastWobbleTargetUpdate: Long = System.nanoTime()

    // Exercise tracking
    @Volatile
    var exerciseStartTime: Long = System.nanoTime()

    @Volatile
    var totalDuration: Double = 0.0

    // Callbacks
    var onPhaseChange: ((BreathPhase, String, Int, Int) -> Unit)? = null
    var onExerciseComplete: ((Int, Double) -> Unit)? = null

    fun reset() {
        state = BreathingExerciseState.STOPPED
        currentPhaseIndex = 0
        currentCycle = 0
        currentScale = 1.0
        targetScale = 1.0
        currentPhase = BreathPhase.IDLE
        currentLabel = ""
        phaseProgress = 0.0
        wobbleIntensity = 1.0
        pauseTime = null
    }
}

File: ios/Breathing/BreathingSharedState.swift
import Foundation

public enum BreathPhase: String {
    case inhale
    case holdIn
    case exhale
    case holdOut
    case idle
}

public enum BreathingExerciseState {
    case stopped
    case running
    case paused
    case complete
}

public final class BreathingSharedState {
    public static let shared = BreathingSharedState()

    // Pattern configuration
    var phases: [BreathPhaseConfig] = []
    var totalCycles: Int? = nil  // nil = infinite

    // Current state
    var state: BreathingExerciseState = .stopped
    var currentPhaseIndex: Int = 0
    var currentCycle: Int = 0
    var phaseStartTime: Date = Date()
    var pauseTime: Date? = nil

    // Animation values (read by TimelineView each frame)
    var currentScale: Double = 1.0
    var targetScale: Double = 1.0
    var startScale: Double = 1.0  // Scale at start of current phase
    var lastScaleUpdate: Date = Date()
    var currentPhase: BreathPhase = .idle
    var currentLabel: String = ""
    var phaseProgress: Double = 0.0  // 0-1 progress through current phase

    // Wobble animation
    var wobbleIntensity: Double = 1.0
    var wobbleOffsets: [Double] = []
    var wobbleTargets: [Double] = []
    var lastWobbleUpdate: Date = Date()
    var lastWobbleTargetUpdate: Date = Date()

    // Exercise tracking
    var exerciseStartTime: Date = Date()
    var totalDuration: Double = 0.0

    // Callbacks
    var onPhaseChange: ((BreathPhase, String, Int, Int) -> Void)? = nil
    var onExerciseComplete: ((Int, Double) -> Void)? = nil

    private init() {}

    func reset() {
        state = .stopped
        currentPhaseIndex = 0
        currentCycle = 0
        currentScale = 1.0
        targetScale = 1.0
        currentPhase = .idle
        currentLabel = ""
        phaseProgress = 0.0
        wobbleIntensity = 1.0
        pauseTime = nil
    }
}

public struct BreathPhaseConfig {
    let phase: BreathPhase
    let duration: Double  // seconds
    let targetScale: Double
    let label: String

    init(phase: BreathPhase, duration: Double, targetScale: Double, label: String) {
        self.phase = phase
        self.duration = duration
        self.targetScale = targetScale
        self.label = label
    }
}
